README.md
# simproxyage: Simulate Age Uncertainty in Climate Proxy Records.

------------------------------

## Introduction

**simproxyage** implements an stochastic approach to simulate the age
uncertainty in _layer-counted_ climate proxy records by perturbing a reference
chronology via randomly omitting or double-counting the reference layers.

The implementation is an adapted and extended version of the MATLAB code
developed by Comboul et al. (2014) who applied it to annually banded coral
archives. The current R package has been used by Münch and Laepple (2018) to
model the time uncertainty of annually dated stable isotope firn-core records
based on seasonal layer counting constrained by volcanic tie points.

The original MATLAB code is available at the [Common Climate Google Code
repository](https://code.google.com/p/common-climate/). The R package version
hosted here has been implemented by Dr. Thomas Münch. Please contact me
<<thomas.muench@awi.de>> at the Alfred Wegener Institute, Helmholtz Centre for
Polar and Marine Research, Germany, for further information.
 
## Installation

**simproxyage** can be installed directly from GitHub:

```r
if (!require("devtools")) {
  install.packages("devtools")
}

devtools::install_github("EarthSystemDiagnostics/simproxyage")
```

## Literature cited

Comboul, M. , Emile-Geay, J., Evans, M. N., Mirnateghi, N., Cobb, K. M. and
Thompson, D. M.: A probabilistic model of chronological errors in layer-counted
climate proxies: applications to annually banded coral archives, Clim. Past,
10(2), 825-841, doi:
[10.5194/cp-10-825-2014](https://doi.org/10.5194/cp-10-825-2014), 2014.

Münch, T. and Laepple, T.: What climate signal is contained in decadal to
centennial scale isotope variations from Antarctic ice cores?, Clim. Past,
accepted for publication, doi:
[10.5194/cp-2018-112](https://doi.org/10.5194/cp-2018-112), 2018.




NEWS.md
# simproxyage 0.1.1

* Installable package version.

# simproxyage 0.1.0

* Package version released along with the publication Münch, T. and Laepple, T.:
_What climate signal is contained in decadal to centennial scale isotope
variations from Antarctic ice cores?_, Clim. Past, 2018.




R/PerturbData.R
##' Age perturb proxy data
##'
##' This function age perturbs input proxy data given on a reference chronology
##' by mapping the proxy values onto a perturbed chronology.
##'
##' The mapping is performed by index matching after rounding of the perturbed
##' age values to integer values (while perturbed ages from \code{\link{BAM}}
##' are integer, non-integer ages can result from tranferring to a Brownian
##' Bridge process (see \code{\link{BrownianBridge}})).
##' @param t.pert numeric vector with the age perturbed chronology to map the
##' proxy data onto.
##' @param t.orig numeric vector with the original reference chronology of the
##' data in \code{x}.
##' @param x numeric vector of the proxy data to perturb.
##' @return Numeric vector with the age perturbed proxy data of the same length
##' as \code{x}.
##' @author Thomas Münch
##' @export
PerturbData <- function(t.pert, t.orig, x) {

    i      <- match(round(t.pert), t.orig)
    x.pert <- x[i]

    return(x.pert)

}



R/BrownianBridge.R
##' Constrained Brownian Bridge process
##'
##' Transfer an unconstrained age perturbation process into a constrained
##' perturbation process following the Brownian Bridge concept, assuming an age
##' control point where the age uncertainty is taken to be zero.
##' @param x numeric vector of the perturbed ages from the unconstrained
##' process; must be of the same length as \code{t}.
##' @param t numeric vector with the original (reference) ages.
##' @param acp the age control point where the age uncertainty is assumed zero.
##' @return Numeric vector of the same length as \code{t} with the perturbed
##' ages following a Brownian Bridge process.
##' @author Thomas Münch
##' @examples
##' t <- 100 : 0
##' ns <- 100
##' bam <- BAM(t, ns = ns)
##' bb <- bam$Tp
##' bb[, 1, ] <- apply(bam$Tp[, 1, , drop = FALSE], c(2, 3), BrownianBridge,
##'                    bam$tp, acp = 1)
##' plot(t, t, type = "n", xlim = c(100, 0), ylim = c(100, 0),
##'      xlab = "Original time", ylab = "Perturbed time")
##' for (i in 1 : ns) {
##' lines(bam$tp, bb[, 1, i], col = "grey", lwd = 1)
##' }
##' abline(0, 1, lwd = 2)

##' @export
BrownianBridge <- function(x, t, acp) {

    t.i   <- seq(t) - 1
    t.acp <- which(t == acp)

    bb <- x - t
    bb <- bb - t.i / (t.acp - 1) * bb[t.acp]
    bb <- bb + t

    return(bb)

}



R/MonteCarloArray.R
##' Monte Carlo simulation of age-perturbed core array
##'
##' For a core array of \code{nc} proxy records, this wrapper function creates
##' \code{nc} identical surrogate time series, independently age-perturbs
##' each series following the model in Comboul et al. (2014) and averages the
##' age-perturbed time series. The process is repeated \code{ns} times. For a
##' common signal recorded by the core array, the approach allows one to study
##' the effect of independent age errors on this signal upon averaging across
##' the core array.
##'
##' Note the principle difference between using \code{PerturbCoreArray} alone
##' and using \code{MonteCarloArray}: while \code{PerturbCoreArray} can also
##' realise \code{ns} age perturbation realisations, these realisations are
##' applied to the same data set in the cores. Here, \code{ns} simulations are
##' run over different surrogate data sets of the cores (however, for the same
##' data in each core) and each simulation run corresponds to one set of age
##' perturbation realisations across the cores. In other words,
##' \code{MonteCarloArray} runs \code{ns} simulations where for each simulation
##' \code{PerturbCoreArray} is called setting the \code{ns} parameter in the
##' function call there to 1.
##' @param t numeric vector of integer values providing a reference chronology
##' for the age perturbations (starting with the youngest age)
##' @param acp numeric vector of age control points where the age uncertainty is
##' assumed to be zero. Per default, a two-element vector where the first
##' element is the start age (\code{t[1]}) and the second element \code{NA}
##' resulting in an unconstrained age perturbation process; see
##' \code{?PerturbCoreArray} for details.
##' @param nt the length of the records (i.e. the number of data points) to
##' simulate; per default set to \code{length(t)}.
##' @param nc the number of cores in the modelled core array
##' @param ns the number of Monte Carlo simulations
##' @inheritParams BAM
##' @param surrogate.fun the random number generator to use for creating the
##' surrogate time series; per default the base \code{R} white noise generator.
##' @param ... additional parameters passed to \code{surrogate.fun}
##' @return a list with components \code{input} and \code{stack}, where both are
##' \code{nt * ns} arrays:
##' \itemize{
##' \item \code{input} contains the \code{ns} realisations of
##' the original surrogate time series
##' \item \code{stack} contains
##' the \code{ns} averages across the \code{nc} individually age-perturbed
##' surrogate time series
##' }
##' @author Thomas Münch
##' @seealso \code{\link{PerturbCoreArray}}
##' @export
MonteCarloArray <- function(t = 100 : 1, acp = c(t[1], NA), nt = length(t),
                            nc = 1, ns = 100, model = "poisson",
                            rate = 0.05, resize = 1,
                            surrogate.fun = rnorm, ...) {

    stacks <- array(data = NA, dim = c(nt, ns))
    input  <- array(data = NA, dim = c(nt, ns))
    
    for (i in 1 : ns) {
        
        # surrogate data to perturb (= nc times the same noise time series)
        X <- array(surrogate.fun(nt, ...), dim = c(nt, nc))
        input[, i] <- X[, 1]

        bam <- PerturbCoreArray(X, t, acp = acp, ns = 1,
                                model = model, rate = rate, resize = resize)

        stacks[, i] <- apply(bam$Xp, 1, mean)

    }

    res <- list()
    res$input <-  input
    res$stacks <- stacks
    
    return(res)

}



R/PerturbCoreArray.R
##' Age perturb proxy core array
##'
##' \code{PerturbCoreArray} age perturbs the proxy data of a similarly dated
##' core array.
##'
##' The function allows one to model the age uncertainty of proxy records with
##' several age control points (ACP) where the age uncertainty evolves according
##' to the model of Comboul et al. (2014) between the age control points but is
##' forced back to zero time uncertainty at the ACPs following a Brownian Bridge
##' process.
##'
##' Per default, the start age (i.e. the youngest age at the core top)
##' is assumed to be the first ACP, thus, if not set explicitly in the vector
##' \code{acp}, the youngest age is added as its first element. You can specify
##' an arbitrary number of additional ACPs. Between each pair of ACPs (starting
##' at the core top), \code{ns} number of constrained age perturbation
##' realisations are performed following the Comboul et al. (2014) model and the
##' Brownian Bridge concept. If the last ACP equals the oldest age in \code{t}
##' (i.e. the last proxy data point), the last core segment also follows a
##' Brownian Bridge process. Alternatively, if the last ACP is younger than the
##' final age, \code{NA} is added as the last element to the vector \code{acp}
##' which results in an unconstrained age perturbation process for the last core
##' segment.
##'
##' ACPs lying outside the time interval defined by \code{t} will be removed
##' from the vector \code{acp} with a warning.
##'
##' The actual age perturbations are performed by calling
##' \code{\link{PerturbCoreSegment}} for each consecutive time interval between
##' two ACPs.
##' @param X an nt * nc array of the proxy data from the core array, where the
##' number of rows (\code{nt}) is the length of the cores (i.e. the length of
##' the proxy data) and the number of columns (\code{nc}) the number of cores in
##' the array.
##' @param t numeric vector providing the original layer-counted
##' chronology. Must equal the length of the proxy data, thus \code{dim(X)[1]}.
##' @param acp numeric vector of age control points where the age uncertainty is
##' assumed to be zero. Per default, a two-element vector where the first
##' element is the start age (\code{t[1]}) and the second element \code{NA}
##' resulting in an unconstrained process; see details for handling more age
##' control points.
##' @inheritParams BAM
##' @return A list of two components:
##' \describe{
##' \item{Tp:}{an nt * nc * ns array of age perturbation realisations for the
##' core array.}
##' \item{Xp:}{an nt * nc * ns array of the corresponding age pertubed proxy
##' data.}
##' }
##'
##' However, the dimension of the arrays depend on the values of \code{nc}
##' (number of cores) and \code{ns} (number of realisations). If one or both
##' values equal 1, the output arrays are converted to the lowest possile
##' dimensionality.
##' @author Thomas Münch
##' @references Comboul, M., Emile-Geay, J., Evans, M. N., Mirnateghi, N., Cobb,
##' K. M. and Thompson, D. M.: A probabilistic model of chronological errors in
##' layer-counted climate proxies: applications to annually banded coral
##' archives, Clim. Past, 10(2), 825-841, doi: 10.5194/cp-10-825-2014, 2014.
##' @seealso \code{\link{BAM}}, \code{\link{BrownianBridge}},
##' \code{\link{PerturbCoreSegment}}, \code{\link{MonteCarloArray}}
##' @export
PerturbCoreArray <- function(X, t, acp = c(t[1], NA), ns = 1000,
                             model = "poisson", rate = 0.05, resize = 1) {

    nt <- dim(X)[1]
    nc <- dim(X)[2]

    if (nt != length(t)) {
        stop("Incompatible lengths of time vector and data.")
    }

    i <- match(acp, t)
    if (!all(!is.na(i))) {
        warning("Some ACPs lie outside of given time vector; will be clipped.")
        acp <- acp[-which(is.na(i))]
    }

    # always treat youngest age as first age control point
    if (acp[1] != t[1]) {
        acp <- c(t[1], acp)
    }

    # treat oldest age as 'NA' age control point
    # if not equal to actual last acp
    if (acp[length(acp)] != t[nt] & !is.na(acp[length(acp)])) {
        acp <- c(acp, NA)
    }

    Xp <- array(data = NA, dim = c(nt, nc, ns))
    Tp <- array(data = NA, dim = c(nt, nc, ns))

    for (i in 1 : (length(acp) - 1)) {

        acp1 <- acp[i]
        acp2 <- acp[i + 1]
        pert <- PerturbCoreSegment(X, t, Xp, Tp, acp1, acp2, ns = ns,
                                   model = model, rate = rate, resize = resize)
        
        Xp <- pert$Xp
        Tp <- pert$Tp

    }

    # simplify arrays to simplest format
    Tp <- drop(Tp)
    Xp <- drop(Xp)

    res <- list()
    res$Tp <- Tp
    res$Xp <- Xp

    return(res)

}

        


R/PerturbCoreSegment.R
##' Age perturb proxy core segments
##'
##' \code{PerturbCoreSegment} age perturbs the proxy data of an individual
##' segment of a similarly dated core array, i.e. the section of the cores that
##' lies between two specified age control points.
##'
##' \code{PerturbCoreSegment}'s intended use is in combination with
##' \code{PerturbCoreArray} only rather than as a standalone function (though
##' possible).
##' @param Xp array of dimension \code{c(dim(X), ns)}, where \code{ns} is the
##' number of age perturbation realisations, to store the perturbed proxy
##' data. To be provided with the calling function.
##' @param Tp array of the same dimension as \code{Xp} to store the age
##' perturbation realisations. To be provided with the calling function.
##' @param acp1 the first age control point, i.e. the start age at the top of
##' the core segment (per default \code{t[1]}).
##' @param acp2 the second age control point at the end of the core segment;
##' \code{NA} (the default) results in an unconstrained age perturbation
##' process.
##' @param nc integer number of cores in the core array; per default set to the
##' number of columns in \code{X}.
##' @inheritParams PerturbCoreArray
##' @return A list of two components:
##' \describe{
##' \item{Tp:}{The array of age perturbation realisations filled with the
##' realisations corresponding to the time interval between \code{acp1} and
##' \code{acp2}.}
##' \item{Xp:}{The array of age perturbed proxy data filled with the
##' perturbation realisations corresponding to the time interval between
##' \code{acp1} and \code{acp2}.}
##' }
##' @author Thomas Münch
##' @references Comboul, M., Emile-Geay, J., Evans, M. N., Mirnateghi, N., Cobb,
##' K. M. and Thompson, D. M.: A probabilistic model of chronological errors in
##' layer-counted climate proxies: applications to annually banded coral
##' archives, Clim. Past, 10(2), 825-841, doi: 10.5194/cp-10-825-2014, 2014.
##' @seealso \code{\link{PerturbCoreArray}}
##' @export
PerturbCoreSegment <- function(X, t, Xp = array(dim = c(dim(X), ns)),
                               Tp = Xp, acp1 = t[1], acp2 = NA, nc = dim(X)[2],
                               ns = 1000, model = "poisson", rate = 0.05,
                               resize = 1) {

    t.final <- acp2
    if (is.na(t.final)) {
        t.final <- t[length(t)]
    }
    
    i <- which(t <= acp1 & t >= t.final)
    n.seg <- length(i)

    t.in <- t[i]
    x.in <- X[i, , drop = FALSE]

    bam <- BAM(t.in, ns = ns, nc = nc, model = model,
               rate = rate, resize = resize)

    bam.Tp <- bam$Tp    
    if (!is.na(acp2)) {

        for (i.c in 1 : nc) {
            
            bam.Tp[, i.c, ] <- apply(bam.Tp[, i.c, , drop = FALSE], c(2, 3),
                                     BrownianBridge, bam$tp, t.final)
        }
    }

    bam.Xp <- array(dim = dim(bam.Tp))
    
    for (i.c in 1 : nc) {

        bam.Xp[, i.c, ] <- apply(bam.Tp[, i.c, , drop = FALSE], c(2, 3),
                                 PerturbData, bam$tp, x.in[, i.c])
    }

    Xp[i, , ] <- bam.Xp[1 : n.seg, , ]
    Tp[i, , ] <- bam.Tp[1 : n.seg, , ]

    res <- list()
    res$Tp <- Tp
    res$Xp <- Xp

    return(res)

}


    


R/BAM.R
##' Generate age perturbations of layer-counted proxy data
##'
##' This function creates an ensemble of perturbed ages of a given reference
##' chronology that is based on layer counting ("banded age model").
##'
##' Ensemble of age perturbations can be realised assuming a single proxy record,
##' or a set of similarly layer-counted proxy data (i.e. a certain number of
##' "cores"). The random age perturbations can either follow a Poisson process
##' or a Bernoulli process; they are realised by randomly removing or doubling
##' time points of the original age model which can result in realisations that
##' are longer (shorter) than the original chronology; in such a case, the
##' output data can be cropped (extended) to match the original length depending
##' on the option specified via \code{resize}. Note that the perturbed ages are
##' automatically flipped to range from most recent to oldest measurements when
##' the input ages are given in increasing order. The code is a modified version
##' of the MATLAB function \code{BAM_simul_perturb.m} provided with the
##' publication  Comboul et al. (2014); please see there for a detailed
##' description of the method.
##' @param t numeric vector providing the original layer-counted chronology.
##' @param ns integer number of age perturbation realisations for each core.
##' @param nc integer number of cores for which an ensemble of age perturbations
##' shall be realised.
##' @param model name string of the random process to use for perturbing the age
##' model; must be either "poisson" (the default) or "bernoulli"; see Comboul et
##' al. (2014) for details on the two models.
##' @param rate numeric vector of probability rate(s) that an age band is
##' perturbed; you can specify a vector of two rates where the first entry is
##' the probability for a missing band and the second entry the probability for
##' a double-counting of a band. If only a single value is specified (per
##' default 0.05), symmetric perturbations are assumed.
##' @param resize the resizing option in case of shorter/longer than original
##' time axes: 0 = do not resize, -1 = resize to shortest realisation, 1 =
##' resize to longest realisation (default).
##' @return A list of two elements:
##' \describe{
##' \item{tp:}{the old chronology cropped/extended depending on the resizing
##' option (thus tp = t for resize = 0).}
##' \item{Tp:}{an array of dimension length(\code{t}) * \code{nc} * \code{ns} with
##' the \code{ns} realisations of perturbed ages for each core \code{nc}.}
##' }
##' @author Maud Comboul; translated and adopted from the original MATLAB code
##' by Thomas Münch
##' @references Comboul, M., Emile-Geay, J., Evans, M. N., Mirnateghi, N., Cobb,
##' K. M. and Thompson, D. M.: A probabilistic model of chronological errors in
##' layer-counted climate proxies: applications to annually banded coral
##' archives, Clim. Past, 10(2), 825-841, doi: 10.5194/cp-10-825-2014, 2014.
##' @examples
##' t <- 100 : 0
##' ns <- 100
##' bam <- BAM(t, ns = ns)
##' plot(t, t, type = "n", xlim = c(100, 0), ylim = c(100, 0),
##'      xlab = "Original time", ylab = "Perturbed time")
##' for (i in 1 : ns) {
##' lines(bam$tp, bam$Tp[, 1, i], col = "grey", lwd = 1)
##' }
##' abline(0, 1, lwd = 2)
##' @export
BAM <- function(t, ns = 1000, nc = 1, model = "poisson",
                rate = 0.05, resize = 1) {

    if (length(t) == 0) {
        stop("Invalid input time vector.")
    }
    
    if (mean(diff(t)) > 0) {
        isflipped <- 1
        t <- rev(t)
    } else {
        isflipped <- 0
    }
    
    n <- length(t)
    p <- nc

    if (length(rate) == 1) {
        rate <- rep(rate, 2)
    }


    # Generate an ensemble of time perturbation models

    time.model <- array(data = 1, dim = c(n, p, ns))
    if (model == "poisson") {
        
        # apply poisson model
        for (nn in 1 : ns) {

            # poisson model for missing bands
            num_event_mis <- rpois(p, rate[1] * n)
            # poisson model for layers counted multiple times
            num_event_dbl <- rpois(p, rate[2] * n)

            for (ii in 1 : p) {
                # place events uniformly on {2,...,n}
                jumps <- sample(n-1, num_event_mis[ii], replace = FALSE) + 1
                # remove 1 at jump locations
                time.model[jumps, ii, nn] <- time.model[jumps, ii, nn] - 1
                # add 1 at jump locations
                jumps <- sample(n-1, num_event_dbl[ii], replace = FALSE) + 1
                time.model[jumps, ii, nn] <- time.model[jumps, ii, nn] + 1
            }
        }

    } else if (model == "bernoulli") {

        # apply bernoulli model
        for (nn in 1 : ns){

            # binomial model for missing bands
            time.model[, , nn] <- time.model[, , nn] -
                array(data = rbinom(n = n * p, size = 1, prob = rate[1]),
                      dim = c(n, p))
            # binomial model for layers counted multiple times
            time.model[, , nn] <- time.model[, , nn] +
                array(data = rbinom(n = n * p, size = 1, prob = rate[2]),
                      dim = c(n, p))
        }

    } else {
        
        stop(paste("Unknown age model;",
                   "acceptable inputs are ''poisson'' and ''bernoulli''."))
    }


    
    # Expand length of time vector if resizing is required
    
    if (resize) {
        
        t_ext <- ceiling(2 * rate[2] * n)
        tn <- n + t_ext

        dt <- t[2] - t[1]
        time_ext <- seq(t[length(t)] + dt, t[length(t)] + t_ext * dt, dt)
        tp <- c(t, time_ext)
        
    } else {
        
        tn <- n
        tp <- t
    }

    
    # Create the pertubed chronologies
    
    Tmax <- 0
    Tmin <- n
    Tp <- array(dim = c(tn, p, ns))
    
    for (nn in 1 : ns) {

        for (ii in 1 : p) {
            xcount <- 1
            Xcount <- 1
            tt <- 1
            while (tt < (n + 1)) {

                if (time.model[tt, ii, nn] == 0) {
                    # remove band
                    Xcount <- min(Xcount + 1, tn)
                } else if (time.model[tt, ii, nn] == 2) {
                    # insert double band
                    Tp[xcount, ii, nn] <- tp[Xcount]
                    xcount <- min(tn, xcount + 1)
                }

                Tp[xcount, ii, nn] <- tp[Xcount]
                xcount <- min(tn, xcount + 1)
                Xcount <- min(tn, Xcount + 1)
                tt <- tt + 1
            }
            
            k <- which(!is.na(Tp[, ii, nn]))
            k <- length(k)
            if (k > Tmax) {
                Tmax <- k
            }
            if (k < Tmin) {
                Tmin <- k
            }
        }
    }


    # Crop output size to shortest sequence
    if (resize == -1) {
        n <- Tmin
    }
    
    # Expand output size to longest (non-NA) sequence
    if (resize == 1) {
        n <- Tmax
    }
    
    Tp <- Tp[1 : n, , , drop = FALSE]
    tp <- tp[1 : n]

    if (isflipped == 1) {
        
        for (nn in 1 : ns) {
            Tp[, , nn] <- apply(Tp[, , nn, drop = FALSE], 2, rev)
        }
        
        tp <- rev(tp)
    }

    res <- list()
    res$tp <- tp
    res$Tp <- Tp
    
    return(res)

}


